content/200-concepts/100-components/02-prisma-client/250-tracing.mdx

```diff
@@ -0,0 +1,213 @@
+
+  You can send tracing output to the console, or analyze it in any OpenTelemetry-compatible tracing system, such as [Jaeger](https://www.jaegertracing.io/).
+
+**TBA: maybe a screenshot in Jaeger with example**
+
+## Output
+
+Depending on the Prisma operation (**request?**), Prisma Client might output any of the following spans:
+
+- `prisma`: Represents the entire Prisma request, from Prisma Client to the database and back. It contains details such as the model and method called by the Prisma Client.
+- `prisma:query_builder`: Represents how long the [Prisma engines](/concepts/components/prisma-engines) take to generate the SQL queries (**would it be more accurate to say "statement or command" instead of queries here? This is not running rhe command, is it? - it's building the query**). This includes deserialization of the Prisma Client request, validation, and transformation into SQL.
+- `prisma:connection`: Represents how long it took for Prisma Client to get a database connection.
+- `prisma:db_query`: Represents a SQL query that was executed against the database, the SQL statement or command (in the tags), and how long the query took to run.
+- `prisma:itx_runner`: The start of an interactive transaction.
+- `prisma:transaction`: The high-level transaction. **need more detail here - what do we mean by high-level?**
+
+## Considerations and prerequisites
+
+**TBA: Garren to provide some notes on performance considerations. We'll say something like "Note that tracing has a significant effect on performance in the following situations". See notes here - https://github.com/prisma/prisma/issues/13503#event-6977115220**
+
+To use tracing, you must do the following:
+
+1. [Install the appropriate dependencies](#step-1-install-up-to-date-prisma-dependencies).
+1. [Enable the `tracing` feature flag in your Prisma schema file](#step-2-enable-the-feature-flag-in-your-prisma-schema-file).
+1. [Configure OpenTelemetry](#step-3-configure-opentelemetry).
+
+### Step 1. Install up-to-date Prisma dependencies
+
+Use version `4.1.0` or higher of the `prisma`, `@prisma/client`, and '@prisma/instrumentation' npm packages.
+
+```terminal
+npm install prisma@latest --save-dev
+npm install @prisma/client@latest --save
+npm install @prisma/instrumentation --save
+```
+
+### Step 2: Enable the feature flag in your Prisma schema file
+
+In the `generator` block of your `schema.prisma` file, enable the `tracing` feature flag:
+
+```prisma
+generator client {
+  provider        = "prisma-client-js"
+  previewFeatures = ["tracing"]
+}
+```
+
+### Step 3: Configure OpenTelemetry
+
+First, install the appropriate OpenTelemetry packages, as follows:
+
+```console
+
+$ npm install @opentelemetry/api @opentelemetry/context-async-hooks @opentelemetry/exporter-trace-otlp-http @opentelemetry/instrumentation @opentelemetry/resources @opentelemetry/sdk-trace-base @opentelemetry/semantic-conventions
+
+```
+
+Then, add the following to your application code:
```

Comment:
This code sample is not quite correct. I've attached a full example and added comments in the sections:

```ts
import { Resource } from '@opentelemetry/resources'
import { SemanticResourceAttributes } from '@opentelemetry/semantic-conventions'
import { BasicTracerProvider, ConsoleSpanExporter, SimpleSpanProcessor } from '@opentelemetry/sdk-trace-base'
import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-http'
import { AsyncHooksContextManager } from '@opentelemetry/context-async-hooks'
import * as api from '@opentelemetry/api'
import { PrismaInstrumentation } from '@prisma/instrumentation'
import { registerInstrumentations } from '@opentelemetry/instrumentation'
import { trace } from '@opentelemetry/api'
import { PrismaClient } from '@prisma/client'

/** SETUP */

export function otelSetup() {
  // a context manager is required to propagate the context
  const contextManager = new AsyncHooksContextManager().enable()

  // it's for node.js for span nesting and ctx propagation
  api.context.setGlobalContextManager(contextManager)

  // a simple exporter that logs the raw data to the console
  const consoleExporter = new ConsoleSpanExporter()

  // exporter that works with jaeger and other OpenTelemetry compliant collector
  const otlpTraceExporter = new OTLPTraceExporter()

  // a standard provider that can run on the web and in node
  const provider = new BasicTracerProvider({
    // sampler: new TraceIdRatioBasedSampler(0.1),
    resource: new Resource({
      // we can define some metadata about the trace resource
      [SemanticResourceAttributes.SERVICE_NAME]: 'sample-tracing-application',
      [SemanticResourceAttributes.SERVICE_VERSION]: '1.0.0',
    }),
  })

  // send spans to the console
  provider.addSpanProcessor(new SimpleSpanProcessor(consoleExporter))
  // send spans to jaeger or any OpenTelemetry compliant collector
  provider.addSpanProcessor(new SimpleSpanProcessor(otlpTraceExporter))

  provider.register()

  // Enable tracing in Prisma
  registerInstrumentations({
    instrumentations: [new PrismaInstrumentation()],
  })
}

//setup tracing for this application
otelSetup()

// set things up for tracing our "app"
const tracer = trace.getTracer('Example Application')

async function main() {
  const prisma = new PrismaClient()

  let users = await tracer.startActiveSpan('create-user', async (span) => {
    try {
      const email = `user.${Date.now()}@sample-application.com`
      await prisma.user.create({
        data: {
          email: email,
        },
      })
      let users = await prisma.user.findMany()
      return users
    } finally {
      span.end()
    }
  })

  console.table(users)
}

main().then(() => console.log('example complete'))



```

Reply:
Thanks, added!